storage/remote/queue_manager.go

```diff
@@ -1058,16 +1060,25 @@ func sendWriteRequestWithBackoff(ctx context.Context, cfg config.QueueConfig, s
 		}
 
 		// If the error is unrecoverable, we should not retry.
-		if _, ok := err.(RecoverableError); !ok {
+		backoffErr, ok := err.(RecoverableError)
+		if !ok {
 			return err
 		}
 
+		sleepDuration = backoff
+		if backoffErr.retryAfter > 0 {
+			sleepDuration = backoffErr.retryAfter
+		}
+		if backoffErr.retryAfter == invalidRetryAfter {
+			level.Debug(l).Log("msg", "retry-after cannot be in past, retrying the default backoff way")
+		}
+
 		// If we make it this far, we've encountered a recoverable error and will retry.
 		onRetry()
 		level.Warn(l).Log("msg", "Failed to send batch, retrying", "err", err)
 
-		time.Sleep(time.Duration(backoff))
-		backoff = backoff * 2
+		time.Sleep(time.Duration(sleepDuration))
```

Comment:
This could block rsharding for hours. I think we need a select that checks for context cancelation here.